### ISSUE TITLE
Hooking up getDocCommentTemplateAtPosition() in Eclipse

### ISSUE THREAD

**Issue by derekcicerone** [Microsoft/TypeScript#4416]
I'm really excited by the new getDocCommentTemplateAtPosition() functionality and I'm trying to hook it up for Eclipse (for https://github.com/palantir/eclipse-typescript).

In Eclipse, the typical way of starting a documentation comment is by typing "/**" and then hitting \<enter>. It looks like the current getDocCommentTemplateAtPosition() may not support this method of autocompleting a doc comment though (but perhaps I'm misunderstanding how to call it).

**Comment by DanielRosenwasser** [comment]
Basically what we do is

1. Determine whether the current location in the buffer is a JSDoc comment completion location.
    1. Basically this consists of
         1. Ensuring the immediately preceding text is `/**` (watch your indexes here!).
         2. Ensuring that the only thing preceding the comment on the current line is whitespace, 
         3. Ensuring *there isn't an existing multiline comment* (basically whether the next non-whitespace is an `*` or EOF. (This is a very rough approximation).
    2. Yes, I think we should move this to the script side, and not do this on the managed side.
2. If the above succeeds, erase the relevant text span for which the `/**` applies, then try to call into the script language service to see if it can give us a JSDoc completion.
3. If we got a completion from the script language service, insert the relevant text.

**Comment by derekcicerone** [comment]
Whoa, ok, thanks for the thorough explanation. I got it working now - this is pretty cool!

I think it might be easier to use the API if the service could handle more of this stuff. In particular, step 2 is kinda tricky because it requires deleting characters, then calling `getDocCommentTemplateAtPosition()`, then restoring the deleted characters if `undefined` is returned (which happens for methods currently because they are unsupported).

**Comment by DanielRosenwasser** [comment]
Yup, I'll discuss with @mhegazy. If @CyrusNajmabadi has any input why this would or would not be a good idea it'd be appreciated.

**Comment by mhegazy** [comment]
The problem is it is a bit late for the language service at this point. ideally you would hold on to the snapshot before the user entered "/**", alternatively, if you are automatically adding the closing comment "*/", you can do the check afterwards.
I agree that the workflow for the API user is not great. @CyrusNajmabadi any suggestions here?

**Comment by CyrusNajmabadi** [comment]
I don't think there is a problem with the API.  I think the right way to think about it "I'm going to ask the question about Doc comments as if there wasn't a doc comment already written in the source.".  This is necessary as the question becomes super difficiult the moment you ask it on code that does have the start of the comment already in there.  i.e. if the code looks like:

```ts
/**
public Blah(i: number) ...
```

Then, from TS's perspective, it just looked like an entirely commented out region of code.

The idea, as Dan said, is that the host will go "let me get the buffer back to the state before the comment was inserted.  Then, i'll ask about doc comments.  Then, if i get something back, i'll insert that in place.  I think that's a non-arduous model to implement.

**Comment by CyrusNajmabadi** [comment]
> Yes, I think we should move this to the script side, and not do this on the managed side.

I don't see a problem with this.  I think we should take the existing Classifier API and move it to a more general "real time, synchronous language service" API.  i.e. it's for features that ask really simple, non-tree, non-syntax, based questions on the text of the document.

Examples of features that would sit here would be:
1. the lexical classifier. It needs to be real time.  It only cares about strings.  
2. the doc comment check.  (i.e. everything we've been talking about).
3. quick questions about completion lists.  i.e. "is this a trigger character".  

Basically, all the places we hardcode smarts about TS into the Managed layer (because we don't want to marshal to another thread when we're on the UI thread) should move into here.  Dan, do you want to do this?  @paulvanbrenk or @vladima could also do it.

**Comment by DanielRosenwasser** [comment]
I thought we were trying to move to the syntactic classifier, why would we invest in the lexical classifier again? It sounds like we'd have to implement a scanner for TypeScript in C#.

Do you mean just a general restructuring of the managed side?

**Comment by CyrusNajmabadi** [comment]
> I thought we were trying to move to the syntactic classifier,

Even with the syntactic classifier, the lexical classifier is still necessary.  

1. It's what is used when a file is open and we don't even have a tree.  If we need ot wait for a tree, you could have black text for several seconds.  (This is easy for you to try out by just disabling the lexical classifier and trying to open a compiler file).

2. Even if we have trees, syntactic classification can be noticeably slow.  This is because perception time for classification is on the order of milliseconds.  Syntactically classifying still involves us having to marshal to another thread, wait for an existing task to finish, marshall over to the script side, have the script side incrementally parse, marshall over the results, and then then have the async tagger display things.  The async tagger itself adds overhead as well as it batches and throttles.  So syntactic classification in the absence of lexical can be very noticeable. 

The benefit of Syntactic classification is *correctness*.  The benefit of Lexical is *perf*.  You want both, and so we need to keep both.  

Note: thanks to @JsonFreeman's great idea about ClassificationTag ordering, we don't need to coordinate the two classifiers.  They can both produce results on their own schedule, and the Syntactic classification tags will naturally override the lexical one.

That way things look really fast (because of the lexical classifier doing things quickly), but they'll always appear accurate almost immediately.  In the 99% case where you're not tying anything lexically odd, you won't even notice because no visible change will happen.

**Comment by CyrusNajmabadi** [comment]
> Do you mean just a general restructuring of the managed side to clarify where these APIs should go?

I mean a restructing on the *script* side.  Right now we have:
1. Lexical operations that happens on the UI thread.  These go through the ClassifierProxy.
2. Syntactic operations that happen on one of two syntactic threads.  
3. Semantic operations that happen on the semantic thread.

On the managed side though we hard-code in several bits of knowledge about TS on the managed side.  This is so we can answer things quickly on the UI thread (since we don't want to jump the syntactic or semantic threads).  These operations should move to the script side.  THe most natural place for this is the Classifier object on the script side.  Except now that will be a terrible name.  Because we'll be asking non-classifier related questions.  So i'm saying, rename that guy to better reflect that it's for *extremely cheap*, *extremely fast*, *text only* questions that the host has.  Use that guy for:

1. lexical classification.
2. quick completion questions. (like "is this a commit character?" or "is this a trigger character?")
3. should i insert a JSDoc comment here?

**Comment by JsonFreeman** [comment]
I think that is very sensible, and I can't imagine there are that many synchronous operations that would have to be moved over. The reason the classifier is its own thread is not because it's the classifier. It's because it happens to be the sole operation that runs synchronously on the UI thread that needs knowledge of TypeScript. That concept could be naturally generalized.

**Comment by mhegazy** [comment]
We should open different issues for the action items from this issue.
